### PR TITLE
The Beta API Channel

### LINKED ISSUE #2509: Implement connectivity API for Python
#2477 introduced the C core API for checking the connectivity state of a channel. This API needs to be wrapped and exposed in any language that exposes a channel-like abstraction.

### PR DESCRIPTION
**Pull request by nathanielmanistaatgoogle**
Fixes #2509.

### CODE REVIEW

File: src/python/grpcio/grpc/_adapter/_c/types/channel.c [hunk 1]
```diff
@@ -164,7 +164,7 @@ PyObject *pygrpc_Channel_watch_connectivity_state(
   int last_observed_state;
   CompletionQueue *completion_queue;
   char *keywords[] = {"last_observed_state", "deadline",
-                      "completion_queue", "tag"};
+                      "completion_queue", "tag", NULL};
```
Comment (soltanmm) [inline src/python/grpcio/grpc/_adapter/_c/types/channel.c]: I thought the format string determined the length of the keywords array. Why is this being added?

File: src/python/grpcio_test/grpc_test/beta/_connectivity_channel_test.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        connectivities = tuple(self._connectivities)
+        if predicate(connectivities):
+          return connectivities
+        else:
+          self._condition.wait()
+
+
+class ChannelConnectivityTest(unittest.TestCase):
+
+  def test_lonely_channel_connectivity(self):
```
Comment (soltanmm) [inline src/python/grpcio_test/grpc_test/beta/_connectivity_channel_test.py]: Not that I feel strongly one way or another, but the majority of test function names we've used so far are camel-case. It's all fine to me - just FYI.

File: src/python/grpcio/grpc/_adapter/_c/types/channel.c (hunk 1 above)
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_adapter/_c/types/channel.c]: Something went wrong this morning, and I looked up the error that I saw in the logs, saw an association with NULL missing from the called function's keyword arguments array, put this NULL in, and didn't see the error any more.

But now that I revert the change, I can't reproduce the error. I'll drop this file from the commit.

File: src/python/grpcio_test/grpc_test/beta/_connectivity_channel_test.py (hunk 2 above)
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_test/grpc_test/beta/_connectivity_channel_test.py]: I *want* them all to be Python-community-compliant snake_case, but I keep falling back on my Google-style and Java-style ingrainedBadHabits. :-( At some point perhaps I'll do a codebase-wide consistency commit?

**Comment by nathanielmanistaatgoogle** [comment]
(Python is green at this commit point but I'll wait for a fresh result after dropping the C file from the commit and repushing.)

File: src/python/grpcio/grpc/_adapter/_c/types/channel.c (hunk 1 above)
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_adapter/_c/types/channel.c]: Okay, I am able to reproduce it, and it seems to be nondeterministic in my primary development environment and deterministically hanging in my secondary development environment. What do you make of "RuntimeError: More keyword list entries (9) than format specifiers (4)" and its relation to the addition of NULL at the end of this array?

Comment (soltanmm) [inline src/python/grpcio/grpc/_adapter/_c/types/channel.c]: ... I am at a complete loss. AFAIK that's undocumented in the C API. I guess I'm NULL-terminating everything from now on.

**Pull request merged**